Adding support for external commits in `referenced` events

Leo-Send opened the pull request:
External commits that mention issues in the project can now be fetched. They receive a new type of related commit, called `commitReferencesIssueExternal`. Also fixed hashatags in markdown code environments incorrectly registering as references to other issues/PRs.

Leo-Send committed:
Add functionality to fetch external commits

external commits that reference an issue can now be fetched. They will
be marked with 'commitReferencesIssueExternal'. To that end, this commit
adds an 'external' field to each commit and a new method to extract
commits using the full url instead of just the relative one within the
repo.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Change 'extractHashtags' to ignore codeblocks

references to issues are now only found if they are ouside a code
environment (starting and ending with three '`'). This mirrors GitHubs
behaviour.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Document findings in 'README.md'

Add paragraph that documents intended and unintended behaviour for
'referenced' events

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueDataProcessor.java:
```diff
@@ -76,7 +76,11 @@ private List<ReferencedLink<GitHubCommit>> parseCommits(IssueData issue) {
                 // filter out errors from referencing commits
                 .filter(eventData -> ((EventData.ReferencedEventData) eventData).commit != null)
-                .map(eventData -> new ReferencedLink<>(Collections.singletonList(((EventData.ReferencedEventData) eventData).commit.getId()), eventData.user, eventData.created_at, "commitReferencesIssue"));
+                .map(eventData -> {if (((GitHubCommit) ((EventData.ReferencedEventData) eventData).commit).getExternal())
+                { return new ReferencedLink<>(Collections.singletonList(((EventData.ReferencedEventData) eventData).commit.getId()), eventData.user, eventData.created_at, "commitReferencesIssueExternal") ;
+            } else {
+                return new ReferencedLink<>(Collections.singletonList(((EventData.ReferencedEventData) eventData).commit.getId()), eventData.user, eventData.created_at, "commitReferencesIssue");
+            }});
 
```
Indentation looks inconsistent to me.

bockthom commented on README.md:
```diff
@@ -43,6 +43,16 @@ java -Xmx100G -jar "build/libs/GitHubWrapper-1.0-SNAPSHOT.jar" \
 
+### `Referenced` events
+
+`Referenced` events are events generated in an issue if a commit references that issue in its commit message. The intended behavior is, that the event is present in the issue's event data, and the commit is again present in the related commits of the issue. This does not work if it is not possible to fetch that commit. In this case, the event still exists, but it contains a link to a commit that  ...
```
"The intended behavior is, that " - no comma before "that".

---

"This may lead to incorrect data points if the resulting data is automatically processed, for example using the tool `codeface-extraction`." This needs more explanation: what are "incorrect data points"? Maybe you could give an example. And if you link to codeface-extraction, then I would link to the issue analysis script there.  Not sure if this is the correct place here to link to it. Maybe we move the section on referenced events to the end and have a separate section before where we link to codeface-extraction and explain that the data could further be processed using this tool? Otherwise a user of the GitHubWrapper might not understand what the relationship with codeface-extraction is and why this does matter...

Leo-Send committed:
Add Section to README.md

reformat README and nested if statement

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on README.md:
```diff
@@ -101,3 +91,19 @@ repo.getIssues(false).ifPresent(issueData -> issueData.forEach(issue -> {
+
+### Further data processing
+
+The data extracted by this tool can be further processed, for example using the `run-issues.py` skript from the tool [`codeface-extraction`](https://github.com/se-sic/codeface-extraction). This organises and unifies the issue data into a single .list file. It also allows for synchronisation with data from other data extraction tools, such as `codeface`.
```
Two comments here:
(1) ... into a single csv-like .list file (nobody knows here what a .list file is, so we should refer to csv)
(2) Could you please use American English here for consistency reasons? means: organizes, synchronization

bockthom commented on README.md:
```diff
@@ -101,3 +91,19 @@ repo.getIssues(false).ifPresent(issueData -> issueData.forEach(issue -> {
+
+The data extracted by this tool can be further processed, for example using the `run-issues.py` skript from the tool [`codeface-extraction`](https://github.com/se-sic/codeface-extraction). This organises and unifies the issue data into a single .list file. It also allows for synchronisation with data from other data extraction tools, such as `codeface`.
+
+### `Referenced` events
+
+`Referenced` events are events generated in an issue if a commit references that issue in its commit message. The intended behavior is that the event is present in the issue's event data, and the commit is again present in the related commits of the issue. This does not work if it is not possible to fetch that commit. In this case, the event still exists, but it contains a link to a commit that t ...
+Known causes of this include:
```
I guess we spell all events with lower-case characters only, don't we? That is, even if we are at the beginning of a sentence / section here, we should spell code items exactly the way they appear in the code.

Leo-Send committed:
Fix spelling in 'README.md'

change BE words for consistency

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Update Copyright headers

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

bockthom commented on src/de/uni_passau/fim/gitwrapper/GitHubCommit.java:
```diff
@@ -119,4 +121,12 @@ public boolean isAddedToPullRequest() {
     }
+
+    void setExternal(boolean external) {
+        this.isExternal = true;
```
Is this a bug or a feature? :thinking:

bockthom commented on README.md:
```diff
@@ -91,3 +91,19 @@ repo.getIssues(false).ifPresent(issueData -> issueData.forEach(issue -> {
+
+### Further data processing
+
+The data extracted by this tool can be further processed, for example using the `run-issues.py` skript from the tool [`codeface-extraction`](https://github.com/se-sic/codeface-extraction). This organizes and unifies the issue data into a single csv-like .list file. It also allows for synchronization with data from other data extraction tools, such as `codeface`.
```
True...

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueDataProcessor.java:
```diff
@@ -262,6 +269,15 @@ private List<String> extractHashtags(String text, boolean onlyInSameRepo) {
+            if (i % 2 == 0) {
+                text = text + texts[i];
+            }
+        }
```
Please check whether this indeed is a performance improvement. Semantics seem to be correct to me.

bockthom commented on src/de/uni_passau/fim/gitwrapper/IssueDataProcessor.java:
```diff
@@ -76,7 +77,13 @@ private List<ReferencedLink<GitHubCommit>> parseCommits(IssueData issue) {
-                .map(eventData -> new ReferencedLink<>(Collections.singletonList(((EventData.ReferencedEventData) eventData).commit.getId()), eventData.user, eventData.created_at, "commitReferencesIssue"));
+                .map(eventData -> {
+                    if (((GitHubCommit) ((EventData.ReferencedEventData) eventData).commit).getExternal()) {
+                        return new ReferencedLink<>(Collections.singletonList(((EventData.ReferencedEventData) eventData).commit.getId()), eventData.user, eventData.created_at, "commitReferencesIssueExternal") ;
```
nitpick but true

bockthom commented on src/de/uni_passau/fim/gitwrapper/GitHubRepository.java:
```diff
@@ -1028,6 +1031,38 @@ Optional<GitHubCommit> getGithubCommit(String hash) {
+                 * JSON String, as it is not needed for further processing.
+                 */
+                LOG.info("Malformed JSON String when querying data for commit " + url + ". Neglect files element.");
+                String jsonStringFromURL = getJSONStringFromURL(url).get();
+                jsonStringFromURL = StringUtils.substringBefore(jsonStringFromURL, "\"files\":[");
+                jsonStringFromURL = jsonStringFromURL + "\"files\":[]}";
+                Optional<GitHubCommit> res = Optional.of(gson.fromJson(jsonStringFromURL, new TypeToken<GitHubCommit>() {}.getType()));
+                checkedHashes.put(hash, res);
+                if (res.isPresent()) {
+                    res.get().setExternal(true);
+                }
+                return res;
```
Please check whether the suggested else case makes sense or not

Leo-Send committed:
Rename field and setter for external commits

Getter previously did not use method parameter. Also renamed and added
docs to match existing methods.

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>

Leo-Send committed:
Improve performance  for extracting hashtags

using stringbuilder instead of appending, which would result in copying
the string

Signed-off-by: Leo Sendelbach <s8lesend@stud.uni-saarland.de>
